[RSDK-10413] create a conan.lock file

penguinland opened the pull request:
This was generated by running `conan lock create .`. I don't like that it doesn't have a newline at the end of the file, but I also don't like manually changing machine-generated files. :shrug: 

The ticket was about investigating whether this is a good idea, but I'm generally in favor of pinning specific versions of dependencies, so I just went ahead and tried it out.

Everything still builds on Linux, though I haven't tried out the Windows or Mac builds. but I've got another ticket to set up a CI, and when that happens, it'll be obvious if there are issues.

penguinland committed:
create conan.lock

penguinland commented on conan.lock:
```diff
@@ -0,0 +1,90 @@
+        "re2/20230301#e8d673155a0494f27c7589b611e8350d%1709641579.83",
+        "protobuf/5.27.0#1fc80ae5667e6cd1d740ca84679b17d6%1742892879.34",
+        "pkgconf/2.2.0#6462942a22803086372db44689ba825f%1713364853.749",
+        "pkgconf/2.1.0#27f44583701117b571307cf5b5fe5605%1701537936.436",
```
Per the discussion in https://github.com/viamrobotics/viam-cpp-sdk/pull/420, multiple versions of the same package are okay in the "build_requires" section because these are things used to build tools and tests, not things to build the final executable. The "requires" section, above, is the stuff that actually goes into the final executable, and each package up there should have one and only one version.

acmorrow commented:
Thanks for doing this. I do have one general question. Once this lockfile is in place, how and when will it get updated? Presumably, that would only be necessary when producing a new release, or upgrading the version of a dependency. Does `conan` have the equivalent of the `--locked` flag in rust tooling, where the build will fail if the lockfile is out of date? That way, if a PR had a change that should result in a lockfile update, but that lockfile update is not made in the PR, the CI run would fail.

acmorrow reviewed (approved):
LGTM but with a mechanics question, and I'm fine with deferring that into a new ticket about how to best work with the lockfile.

penguinland committed:
make module.tar.gz should work, and the conan build artifacts should not be tracked

penguinland commented:
> Once this lockfile is in place, how and when will it get updated? 

When you want to change one of the dependencies, you update this file, too. `conan lock --help` will explain more, but you can add, remove, and update requirements using the CLI.

> the build will fail if the lockfile is out of date?

Yes. As an example, I changed the "requires" section to include `"protobuf/3.21.12#d9f5f4e3b86552552dda4c0a2e928eeb%1685218275.69"` instead of version 5.27.0, and running `make tflite_cpu` failed with `ERROR: Requirement 'protobuf/5.27.0' not in lockfile 'requires'`. Undoing that change to conan.lock got the build to succeed.

(While checking on that, I had errors that `bin/package.sh` was not executable, and I got annoyed that `build-conan/` made my repo dirty. So, I've changed both of those things in this PR, too)

acmorrow commented:
> > Once this lockfile is in place, how and when will it get updated?
> 
> When you want to change one of the dependencies, you update this file, too. `conan lock --help` will explain more, but you can add, remove, and update requirements using the CLI.
> 
> > the build will fail if the lockfile is out of date?
> 
> Yes. As an example, I changed the "requires" section to include `"protobuf/3.21.12#d9f5f4e3b86552552dda4c0a2e928eeb%1685218275.69"` instead of version 5.27.0, and running `make tflite_cpu` failed with `ERROR: Requirement 'protobuf/5.27.0' not in lockfile 'requires'`. Undoing that change to conan.lock got the build to succeed.
> 

Sounds good, as long as you as the maintainer are happy with the semantics and workflow then I am too.


> (While checking on that, I had errors that `bin/package.sh` was not executable, and I got annoyed that `build-conan/` made my repo dirty. So, I've changed both of those things in this PR, too)

Yeah, to be entirely honest I wasn't quite sure what the deal was with putting things inside a `build-conan` directory rather than just using the regular old `build` that I assume would be the default? I just followed along with it because I was short on time and didn't want to lose focus investigating. But personally I'd be all for just eliminating the `build-conan` thing.

penguinland commented:
:shrug: I'm fine with `build-conan` or `build`, and it's already working as-is. Feel free to rename if you want, but my hope is that I can ignore whatever it is.
